soh/soh/Enhancements/controls/GameControlEditor.cpp

```diff
@@ -216,36 +215,132 @@ namespace GameControlEditor {
         }
         
         ImVec2 cursor = ImGui::GetCursorPos();
         ImGui::SetCursorPos(ImVec2(cursor.x + 5, cursor.y + 5));
         SohImGui::PaddedEnhancementCheckbox("Invert Camera X Axis", "gInvertXAxis");
-		SohImGui::Tooltip("Inverts the Camera X Axis in:\n-Free camera\n-C-Up view\n-Weapon Aiming");
-		ImGui::SetCursorPosX(ImGui::GetCursorPosX() + 5);
-		SohImGui::PaddedEnhancementCheckbox("Invert Camera Y Axis", "gInvertYAxis");
-		SohImGui::Tooltip("Inverts the Camera Y Axis in:\n-Free camera\n-C-Up view\n-Weapon Aiming");
-		ImGui::SetCursorPosX(ImGui::GetCursorPosX() + 5);
-		SohImGui::PaddedEnhancementCheckbox("Right Stick Aiming", "gRightStickAiming");
-		SohImGui::Tooltip("Allows for aiming with the rights stick when:\n-Aiming in the C-Up view\n-Aiming with weapons");
-		ImGui::SetCursorPosX(ImGui::GetCursorPosX() + 5);
-		SohImGui::PaddedEnhancementCheckbox("Auto-Center First Person View", "gAutoCenterView");
-		SohImGui::Tooltip("Prevents the C-Up view from auto-centering, allowing for Gyro Aiming");
-	}
-	
-	void DrawUI(bool& open) {
+        DrawHelpIcon("Inverts the Camera X Axis in:\n-Free camera\n-C-Up view\n-Weapon Aiming");
+        ImGui::SetCursorPosX(ImGui::GetCursorPosX() + 5);
+        SohImGui::PaddedEnhancementCheckbox("Invert Camera Y Axis", "gInvertYAxis");
+        DrawHelpIcon("Inverts the Camera Y Axis in:\n-Free camera\n-C-Up view\n-Weapon Aiming");
+        ImGui::SetCursorPosX(ImGui::GetCursorPosX() + 5);
+        SohImGui::PaddedEnhancementCheckbox("Right Stick Aiming", "gRightStickAiming");
+        DrawHelpIcon("Allows for aiming with the rights stick when:\n-Aiming in the C-Up view\n-Aiming with weapons");
+        ImGui::SetCursorPosX(ImGui::GetCursorPosX() + 5);
+        SohImGui::PaddedEnhancementCheckbox("Auto-Center First Person View", "gAutoCenterView");
+        DrawHelpIcon("Prevents the C-Up view from auto-centering, allowing for Gyro Aiming");
+    }
+    
+    void DrawUI(bool& open) {
         if (!open) {
             CVar_SetS32("gGameControlEditorEnabled", false);
             return;
         }
 
         ImGui::SetNextWindowSize(ImVec2(465, 430), ImGuiCond_FirstUseEver);
         if (ImGui::Begin("Game Controls Configuration", &open)) {
-            DrawOcarinaControlPanel();
-			DrawCameraControlPanel();
+            ImGui::BeginTabBar("##CustomControllers");
+            if (ImGui::BeginTabItem("Generic")) {
+                CurrentPort = 0;
+                ImGui::EndTabItem();
+            }
+
+            for (int i = 1; i <= 4; i++) {
+                if (ImGui::BeginTabItem(StringHelper::Sprintf("Port %d", i).c_str())) {
+                    CurrentPort = i;
+                    ImGui::EndTabItem();
+                }
+            }
+
+            ImGui::EndTabBar();
+
+            if (CurrentPort == 0) {
+                DrawOcarinaControlPanel();
+                DrawCameraControlPanel();
+            } else {
+                DrawCustomButtons();
+            }
```

Comment:
only drawing custom buttons on ports 2-4?

Reply:
No, I had to use different indices than what are used in the Input Editor because we have a "general" tab for controls that aren't controller-specific.  Now, 0 is the general tab and 1 - 4 actually correspond to ports 1 - 4.  You'll notice that wherever I use the CurrentPort variable, I always have (CurrentPort - 1), as opposed to its corresponding usage in the Input Editor.

Reply:
ah, yeah, that makes sense, good deal